katpoint/projection.py

```diff
@@ -129,24 +133,196 @@
+# --- Handling out-of-range inputs
+# --------------------------------------------------------------------------------------------------
+
+
+class OutOfRangeError(ValueError):
+    """A numeric value is out of range."""
+
+
+class OutOfRange(object):
+    """Customisable handling of out-of-range numeric values."""
+
+    _treatment = _old_treatment = 'raise'
+
+    @classmethod
+    def __enter__(cls):
+        pass
+
+    @classmethod
+    def __exit__(cls, *args, **kws):
+        cls._treatment = cls._old_treatment
+
+    @classmethod
+    def get_treatment(cls):
+        """The current treatment of out-of-range values."""
+        return cls._treatment
+
+    @classmethod
+    def set_treatment(cls, treatment):
+        """"Change the treatment of out-of-range values.
+
+        The supported treatments are:
+          - 'raise': raise :class:`OutOfRangeError` (the default)
+          - 'nan': replace out-of-range values with NaNs
+          - 'clip': replace out-of-range values with nearest valid values
+
+        Parameters
+        ----------
+        treatment : {'raise', 'nan', 'clip'}
+            New treatment
+
+        Returns
+        -------
+        context_manager : :class:`OutOfRange`
+            An instance intended to be used in a with-statement
+
+        Raises
+        ------
+        ValueError
+            If `treatment` is not a recognised option
+        """
+        valid_treatments = {'raise', 'nan', 'clip'}
+        if treatment not in valid_treatments:
+            raise ValueError("Unknown out-of-range treatment '{}', must be one of {}"
+                             .format(treatment, valid_treatments))
+        cls._old_treatment = cls._treatment
+        cls._treatment = treatment
+        return cls()
+
+    @classmethod
+    def treat(cls, x, err_msg, lower=None, upper=None):
+        """Apply treatment to any out-of-range values in `x`.
+
+        Parameters
+        ----------
+        x : real number or array-like of real numbers
+            Input values (left untouched)
+        err_msg : string
+            Error message passed to exception if treatment is 'raise'
+        lower, upper : real number or None, optional
+            Bounds for values in `x` (specify at least one bound!)
+
+        Returns
+        -------
+        treated_x : float or array of float
+            Treated values (guaranteed to be in range or NaN)
+
+        Raises
+        ------
+        OutOfRangeError
+            If any values in `x` are out of range and treatment is 'raise'
+
+        Notes
+        -----
+        If a value is out of bounds by less than the machine precision, it is
+        considered a rounding error. It is not treated as out-of-range to avoid
+        false alarms, but instead silently clipped to ensure that all returned
+        data is in the valid range.
+        """
+        clipped_x = np.asarray(np.clip(x, lower, upper), dtype=np.float)
```

Comment:
Also just saw that `np.float` is going to be [deprecated](https://numpy.org/devdocs/release/1.20.0-notes.html#using-the-aliases-of-builtin-types-like-np-int-is-deprecated) in the next version of numpy:

Reply:
I did that back in April without documenting it, but I suspect it is so that I can do
```Python
clipped_x[out_of_range] = np.nan
```
I'll make a comment and fix the `float`.

Reply:
settle for 'clip' option for now.